machine-health-checks.md

```diff
@@ -0,0 +1,117 @@
+  workers:
+    machineDeployments:
+    - class: default-worker
+      template:
+      ...
+      machineHealthCheck:
+        unhealthyConditions:
+        - type: DemoNodeHealthy
+          status: "False"
+          timeout: 60s
+```
+
+Based on this configuration Cluster API created a MachineHealthCheck object for our MachineDeployment:
+
+```bash
+kubectl get machinedeployment,machinehealthcheck
+```
+
+Output:
+
+```bash
+NAME                                                               CLUSTER              REPLICAS   READY   UPDATED   UNAVAILABLE   PHASE       AGE     VERSION
+machinedeployment.cluster.x-k8s.io/docker-cluster-one-md-0-np68x   docker-cluster-one   1                  1         1             ScalingUp   4m42s   v1.24.6
+
+NAME                                                                CLUSTER              EXPECTEDMACHINES   MAXUNHEALTHY   CURRENTHEALTHY   AGE
+machinehealthcheck.cluster.x-k8s.io/docker-cluster-one-md-0-np68x   docker-cluster-one   1                  100%           1                4m42s
+```
+
+The MachineHealthCheck will now monitor the Nodes of the MachineDeployment. As soon as the `DemoNodeHealthy` condition is detected with the value `False` 
+the MachineHealthCheck will mark the Machine for remediation via the `OwnerRemediated` condition set to `False`.
+
+This will indicate to the MachineSet controller (which is the owner of the Machine) that the Machine must be remediated. Thus the MachineSet controller will drain and delete the Machine.
+After that the MachineSet controller will create a new Machine to enxure we still have as many replicas as specified in the MachineSet `spec.replicas` field.
+
+Now let's first take a look at our current Machines:
+
+```bash
+kubectl get machine
+```
+
+Output:
+```bash
+kubectl get machine
+NAME                                             CLUSTER              NODENAME                                         PROVIDERID                                                  PHASE     AGE     VERSION
+docker-cluster-one-md-0-np68x-6f47ffdffb-zqvv7   docker-cluster-one   docker-cluster-one-md-0-np68x-6f47ffdffb-zqvv7   docker:////docker-cluster-one-md-0-np68x-6f47ffdffb-zqvv7   Running   3m40s   v1.24.6
+docker-cluster-one-pkkxd-66wmm                   docker-cluster-one   docker-cluster-one-pkkxd-66wmm                   docker:////docker-cluster-one-pkkxd-66wmm                   Running   24m     v1.24.6
+```
+
+And our current Nodes:
+
+```bash
+kubectl --kubeconfig cluster-one.kubeconfig get node
+```
+
+Output:
+```bash
+NAME                                             STATUS   ROLES           AGE     VERSION
+docker-cluster-one-md-0-np68x-6f47ffdffb-zqvv7   Ready    <none>          3m46s   v1.24.6
+docker-cluster-one-pkkxd-66wmm                   Ready    control-plane   24m     v1.24.6
+```
+
+Then let's simulate a Node failure by setting the `DemoNodeHealthy` condition to `False`: 
+
+**Note**: You have to use the Node name of the `md-1` node as returned above.
```

Comment:
I wonder if that's really easier then just telling people to lookup the node name.

I men we can add a script which does a node list + grep md -1  + patch the node.

But I wonder if they understand more what's going on by doing it directly

Reply:
I think if we explain in the text what's happening and make the commands sort-of magic that's a smoother experience for the majority of participants. Most of the other commands (I think it should be all) work just copy-pasting, and I think we should try to keep to that flow.

This should work I think:

```
kubectl --kubeconfig cluster-one.kubeconfig patch node $(kubectl --kubeconfig cluster-one.kubeconfig get nodes -l 'node-role.kubernetes.io/control-plane notin () ' -o jsonpath={'.items[*].metadata.name'}) --subresource=status --type=json -p='[{"op": "add", "path": "/status/conditions/-", "value": {"type": "DemoNodeHealthy", "status": "False", "message": "Node is unhealthy"}}]'
```

Reply:
I'll give it a try. I just hope we don't have to debug this statement for folks

Reply:
Hm I think this will fail if they still have more than 1 MD repica

Reply:
Or maybe it will flag all of them? IIRC the MHC controller then blocks because too many Machines are unhealthy (had something like this when exploring cert rotation)

Reply:
Okay tested. It flags all worker nodes as unhealthy + we get a remediation. That's fine

Reply:
We could also index the command i.e. 

```
kubectl --kubeconfig cluster-one.kubeconfig patch node $(kubectl --kubeconfig cluster-one.kubeconfig get nodes -l 'node-role.kubernetes.io/control-plane notin () ' -o jsonpath={'.items[*].metadata.name'}) --subresource=status --type=json -p='[{"op": "add", "path": "/status/conditions/-", "value": {"type": "DemoNodeHealthy", "status": "False", "message": "Node is unhealthy"}}]'
```